### PR TITLE
Add Grid bulk action methods

### LINKED ISSUE #1727: Execute php callback on Grid->addSelection() for multiple selection
First of all, I want to thank you for this great tool that is atk4/ui !
Since some time I have a question which I believe does not find answer in the documentation (but maybe I am just blinded).
Let's suppose we have two models `User` and `Role` and `User` `hasMany` `Role`.

- It is quite straightforward to use `\Atk4\Ui\Grid` and `addCondition` to display a given `Role` and add a `User`.
- Also we can almost out of the box display the `Role` list of a `User` and add one using the same method.

My question is : how to use `$grid->addSelection()` to mass-attribute a `Role` to a selected set of `User` ? (Which enable to use grid filtering and sorting)
In the [demo](https://github.com/atk4/ui/blob/develop/demos/collection/grid.php) there is a call to some javascript
```
$grid->menu->addItem('show selection')->on('click', new \Atk4\Ui\JsExpression(
    'alert("Selected: "+[])',
    [$sel->jsChecked()]
));
```
But I does not understand how to relies on a php callback instead that would receive the selected id list and perform some processing such at foreach the list and create the appropriate role lines or others.

### PR DESCRIPTION
**Pull request by mhuser**
fix #1727

This PR aims to add the bulk action feature to process a selection or multiple records in Grid. Currently it is easy to perform an operation:
1.  on a single record e.g. with `addModalAction()`
2. on the whole grid e.g. with `$grid->menu->addItem()`)

It is complex to work with the results of `$grid->addSelection()` outside writing verbose code with static modals and a hidden field.

### CODE REVIEW

File: demos/collection/grid.php [hunk 1]
```diff
@@ -72,6 +74,45 @@
 // $grid->addExecutorButton($deleteExecutor, new Button(['icon' => 'times circle outline']));
 
 $sel = $grid->addSelection();
+// Executing a modal on a bulk selection
+$callback = function ($m, $ids) use ($grid) {
```
Comment (mvorisek) [inline demos/collection/grid.php]: types should be defined, `$m` variable should be named `$modal`

File: demos/collection/grid.php [hunk 2]
```diff
@@ -72,6 +74,45 @@
...
+    if (!$ids) {
+        Message::addTo($m, [
```
Comment (mvorisek) [inline demos/collection/grid.php]: Build action button should be disabled instead when nothing is selected.

File: src/Grid.php [hunk 3]
```diff
@@ -513,6 +513,37 @@ public function addModalAction($button, $title, \Closure $callback, $args = [])
...
+     * @param string|array|MenuItem $item
+     * @param \Closure              $callback function (View $page) {...
+     * @param array                 $args     extra url argument for callback
+     *
+     * @return Jquery
+     */
+    public function addModalBulkAction($item, \Closure $callback, $args = [])
+    {
+        if (!$this->menu) {
+            throw new Exception('Unable to add Modal Bulk Action without Menu');
```
Comment (mvorisek) [inline src/Grid.php]: explicit exception is not needed

File: demos/collection/grid.php [hunk 4]
```diff
@@ -72,6 +74,45 @@
...
+        $msg->text->addParagraph('IDs to be deleted:');
+        foreach ($ids as $id) {
+            $msg->text->addParagraph($id);
+        }
+        $form = Form::addTo($m);
+        $form->buttonSave->set('Delete');
+        $form->buttonSave->icon = 'trash';
+        $form->onSubmit(function (Form $form) use ($grid, $ids) {
+            // iterate trough the selected IDs and delete them
+            $grid->model->atomic(function () use ($grid, $ids) {
```
Comment (mvorisek) [inline demos/collection/grid.php]: Must be covered with Behat test.

File: src/Grid.php [hunk 5]
```diff
@@ -513,6 +513,37 @@ public function addModalAction($button, $title, \Closure $callback, $args = [])
...
+        }
+
+        $owner = $this->getOwner();
+
+        if (is_string($item)) {
+            $item = ['title' => $item];
+        }
+
+        $modal = Modal::addTo($owner, [$item[0]]);
```
Comment (mvorisek) [inline src/Grid.php]: `$item[0]` cannot be correct here, at least not always set

Comment (mvorisek) [inline src/Grid.php]: no need to assing to `$owner` first, use `$this->getOwner()` directly

File: src/Grid.php [hunk 6]
```diff
@@ -513,6 +513,37 @@ public function addModalAction($button, $title, \Closure $callback, $args = [])
         return $this->getActionButtons()->addModal($button, $title, $callback, $this, $args);
     }
 
+    /**
+     * Similar to addModalAction but apply to a multiple recors selection and display in menu.
+     * When menu item is clicked, modal is displayed with the $title and $callback is executed through VirtualPage.
+     *
+     * @param string|array|MenuItem $item
+     * @param \Closure              $callback function (View $page) {...
```
Comment (mvorisek) [inline src/Grid.php]: document `\Closure(XX): XX` type better

File: demos/collection/grid.php (hunk 1 above)
Comment (mhuser) [reply demos/collection/grid.php]: done

File: demos/collection/grid.php (hunk 2 above)
Comment (mhuser) [reply demos/collection/grid.php]: nice idea. have you an example of such magic?

File: src/Grid.php (hunk 3 above)
Comment (mhuser) [reply src/Grid.php]: done

File: src/Grid.php (hunk 5 above)
Comment (mhuser) [reply src/Grid.php]: it appear it is not needed at all

Comment (mhuser) [reply src/Grid.php]: ok, done

File: src/Grid.php (hunk 6 above)
Comment (mhuser) [reply src/Grid.php]: done

File: demos/collection/grid.php (hunk 2 above)
Comment (mhuser) [reply demos/collection/grid.php]: Maybe the master checkbox code could enable/disable bulk action button by querying a dedicated class

File: src/Grid.php [hunk 7]
```diff
@@ -513,6 +513,32 @@ public function addModalAction($button, $title, \Closure $callback, $args = [])
...
     }
 
+    /**
+     * Similar to addModalAction but apply to a multiple recors selection and display in menu.
+     * When menu item is clicked, modal is displayed with the $title and $callback is executed through
+     * VirtualPage.
+     *
+     * @param string|array|MenuItem $item
+     * @param \Closure(View, array|null): void $callback
+     * @param array                            $args     extra url argument for callback
```
Comment (mvorisek) [inline src/Grid.php]:
```suggestion
     * @param string|array|MenuItem              $item
     * @param \Closure(View, list<string>): void $callback
     * @param array                              $args     extra url argument for callback
```

`$args` are probably not needed - what is the usecase?

also I do not understand why explicit `[]` is needed as `$this->menu->addItem` 2nd arg

File: src/Grid.php [hunk 8]
```diff
@@ -513,6 +513,32 @@ public function addModalAction($button, $title, \Closure $callback, $args = [])
...
+     * When menu item is clicked, modal is displayed with the $title and $callback is executed through
+     * VirtualPage.
+     *
+     * @param string|array|MenuItem            $item
+     * @param \Closure(View, array|null): void $callback
+     * @param array                            $args     extra url argument for callback
+     *
+     * @return Jquery
+     */
+    public function addModalBulkAction($item, \Closure $callback, $args = [])
```
Comment (mvorisek) [inline src/Grid.php:527]:
```suggestion
    public function addModalBulkAction($item, \Closure $callback, $args = []): JsExpressionable
```

File: src/Grid.php [hunk 9]
```diff
@@ -513,6 +513,32 @@ public function addModalAction($button, $title, \Closure $callback, $args = [])
...
+            $item = ['title' => $item];
+        }
+
+        $modal = Modal::addTo($this->getOwner());
+
+        $modal->set(function (View $t) use ($callback) {
+            $callback($t, $t->stickyGet($this->name) ? explode(',', $t->stickyGet($this->name)) : null);
+        });
+
+        return $this->menu->addItem($item)->on('click', $modal->jsShow(array_merge([$this->name => $this->selection->jsChecked()], $args)), []);
```
Comment (mvorisek) [inline src/Grid.php]: any reason why this return JsExpressionable and non-bulk method View?

File: src/Grid.php [hunk 10]
```diff
@@ -513,6 +513,32 @@ public function addModalAction($button, $title, \Closure $callback, $args = [])
         return $this->getActionButtons()->addModal($button, $title, $callback, $this, $args);
     }
 
+    /**
+     * Similar to addModalAction but apply to a multiple recors selection and display in menu.
+     * When menu item is clicked, modal is displayed with the $title and $callback is executed through
+     * VirtualPage.
+     *
+     * @param string|array|MenuItem            $item
+     * @param \Closure(View, array|null): void $callback
```
Comment (mvorisek) [inline src/Grid.php]: why to pass `null` when `$t->stickyGet($this->name)` is empty?

Comment (mvorisek) [reply src/Grid.php]: fix this, add a Behat test and I will do a final review

no typecasting is ok as non-bulk action does not have it now too, will be fixed later in #1991

File: src/Grid.php [hunk 11]
```diff
@@ -513,6 +513,32 @@ public function addModalAction($button, $title, \Closure $callback, $args = [])
         return $this->getActionButtons()->addModal($button, $title, $callback, $this, $args);
     }
 
+    /**
+     * Similar to addModalAction but apply to a multiple recors selection and display in menu.
```
Comment (mvorisek) [inline src/Grid.php]:
```suggestion
     * Similar to addModalAction but apply to a multiple records selection and display in menu.
```

File: demos/collection/grid.php (hunk 2 above)
Comment (mhuser) [reply demos/collection/grid.php]: Wait on #1921 in the meantime, the action will return a void array, then iterating through it will do nothing.

**Comment by mhuser** [comment]
Maybe there is an interest to `addBulkAction` too, that would fire without `Modal`.

File: demos/collection/grid.php (hunk 4 above)
Comment (mhuser) [reply demos/collection/grid.php]: Done

File: src/Grid.php (hunk 7 above)
Comment (mhuser) [reply src/Grid.php]: `[]` removed; not sure for `$args` but `addModalAction` has this and I used it as template.

File: src/Grid.php (hunk 9 above)
Comment (mhuser) [reply src/Grid.php]: Should return View.

Comment (mhuser) [reply src/Grid.php]: Done

File: src/Grid.php (hunk 10 above)
Comment (mhuser) [reply src/Grid.php]: Replaced null by empty array and added behat test

**Comment by mvorisek** [comment]
Quality PR, thank you!

**Comment by mhuser** [comment]
> Quality PR, thank you!

Great !

**Pull request merged**